From 30b806e95404622c7e82845925dcfc8680045319 Mon Sep 17 00:00:00 2001
From: Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>
Date: Tue, 24 Sep 2024 16:46:36 +0530
Subject: [PATCH] PENDING: allow systemd generator to getattr of some unit and
 script files

allow systemd to getattr of :
		avahi unit and script files
		bluetooth unit and script files
		chronyd unit and script files
		alsa unit file
		rpcbind script file
		rpcd unit file
		systemd unit file

Upstream-Status: Pending

Signed-off-by: Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>
---
 policy/modules/admin/alsa.if         | 19 +++++++++++++
 policy/modules/services/avahi.if     | 38 ++++++++++++++++++++++++++
 policy/modules/services/bluetooth.if | 40 ++++++++++++++++++++++++++++
 policy/modules/services/chronyd.if   | 38 ++++++++++++++++++++++++++
 policy/modules/services/rpc.if       | 19 +++++++++++++
 policy/modules/services/rpcbind.if   | 19 +++++++++++++
 policy/modules/system/systemd.te     | 11 +++++++-
 7 files changed, 183 insertions(+), 1 deletion(-)

diff --git a/policy/modules/admin/alsa.if b/policy/modules/admin/alsa.if
index 57dd48df4..b1b0bed06 100644
--- a/policy/modules/admin/alsa.if
+++ b/policy/modules/admin/alsa.if
@@ -45,6 +45,25 @@ interface(`alsa_run',`
 	roleattribute $2 alsa_roles;
 ')
 
+########################################
+## <summary>
+##   Getattr for the alsa init unit file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr alsa
+##      init unit file.
+##      </summary>
+## </param>
+#
+interface(`alsa_getattr_init_unit_file',`
+        gen_require(`
+                type alsa_unit_t;
+        ')
+
+        allow $1 alsa_unit_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	Read and write Alsa semaphores.
diff --git a/policy/modules/services/avahi.if b/policy/modules/services/avahi.if
index fe9f460e3..281bc9156 100644
--- a/policy/modules/services/avahi.if
+++ b/policy/modules/services/avahi.if
@@ -38,6 +38,44 @@ interface(`avahi_initrc_domtrans',`
 	init_labeled_script_domtrans($1, avahi_initrc_exec_t)
 ')
 
+########################################
+## <summary>
+##   Getattr for the avahi init unit file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr avahi
+##      init unit file.
+##      </summary>
+## </param>
+#
+interface(`avahi_getattr_init_unit_file',`
+        gen_require(`
+                type avahi_unit_t;
+        ')
+
+        allow $1 avahi_unit_t:file getattr;
+')
+
+########################################
+## <summary>
+##   Getattr for the avahi init script file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr of avahi
+##      init script file.
+##      </summary>
+## </param>
+#
+interface(`avahi_getattr_init_script_file',`
+        gen_require(`
+                type avahi_initrc_exec_t;
+        ')
+
+        allow $1 avahi_initrc_exec_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	Send generic signals to avahi.
diff --git a/policy/modules/services/bluetooth.if b/policy/modules/services/bluetooth.if
index bd3dff2ff..0fb15c6eb 100644
--- a/policy/modules/services/bluetooth.if
+++ b/policy/modules/services/bluetooth.if
@@ -109,6 +109,46 @@ interface(`bluetooth_use',`
 	bluetooth_stream_connect($1)
 ')
 
+########################################
+## <summary>
+##   Getattr for the bluetooth
+##   init unit file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr
+##      bluetooth init unit file.
+##      </summary>
+## </param>
+#
+interface(`bluetooth_getattr_init_unit_file',`
+        gen_require(`
+                type bluetooth_unit_t;
+        ')
+
+        allow $1 bluetooth_unit_t:file getattr;
+')
+
+########################################
+## <summary>
+##   Getattr for the bluetooth init
+##   script file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr of
+##      bluetooth init script file.
+##      </summary>
+## </param>
+#
+interface(`bluetooth_getattr_init_script_file',`
+        gen_require(`
+                type bluetooth_initrc_exec_t;
+        ')
+
+        allow $1 bluetooth_initrc_exec_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	Execute bluetooth in the bluetooth domain.
diff --git a/policy/modules/services/chronyd.if b/policy/modules/services/chronyd.if
index 4c26a68c4..0232025fd 100644
--- a/policy/modules/services/chronyd.if
+++ b/policy/modules/services/chronyd.if
@@ -38,6 +38,44 @@ interface(`chronyd_domtrans_cli',`
 	domtrans_pattern($1, chronyc_exec_t, chronyc_t)
 ')
 
+########################################
+## <summary>
+##   Getattr for chronyd init unit file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr chronyd
+##      init unit file.
+##      </summary>
+## </param>
+#
+interface(`chronyd_getattr_init_unit_file',`
+        gen_require(`
+                type chronyd_unit_t;
+        ')
+
+        allow $1 chronyd_unit_t:file getattr;
+')
+
+########################################
+## <summary>
+##   Getattr for the chronyd init script file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr of chronyd
+##      init script file.
+##      </summary>
+## </param>
+#
+interface(`chronyd_getattr_init_script_file',`
+        gen_require(`
+                type chronyd_initrc_exec_t;
+        ')
+
+        allow $1 chronyd_initrc_exec_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	Execute chronyd server in the
diff --git a/policy/modules/services/rpc.if b/policy/modules/services/rpc.if
index e47728196..fe08d3ea4 100644
--- a/policy/modules/services/rpc.if
+++ b/policy/modules/services/rpc.if
@@ -50,6 +50,25 @@ template(`rpc_domain_template',`
 	auth_use_nsswitch($1_t)
 ')
 
+########################################
+## <summary>
+##   Getattr for the rpcd init unit file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr rpcd
+##      init unit file.
+##      </summary>
+## </param>
+#
+interface(`rpcd_getattr_init_unit_file',`
+        gen_require(`
+                type rpcd_unit_t;
+        ')
+
+        allow $1 rpcd_unit_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	List export files.
diff --git a/policy/modules/services/rpcbind.if b/policy/modules/services/rpcbind.if
index 0938487d2..c72fe6633 100644
--- a/policy/modules/services/rpcbind.if
+++ b/policy/modules/services/rpcbind.if
@@ -39,6 +39,25 @@ interface(`rpcbind_stream_connect',`
 	stream_connect_pattern($1, rpcbind_runtime_t, rpcbind_runtime_t, rpcbind_t)
 ')
 
+########################################
+## <summary>
+##   Getattr for the rpcbind init script file.
+## </summary>
+## <param name="type">
+##      <summary>
+##      Type to be used for getattr of rpcbind
+##      init script file.
+##      </summary>
+## </param>
+#
+interface(`rpcbind_getattr_init_script_file',`
+        gen_require(`
+                type rpcbind_initrc_exec_t;
+        ')
+
+        allow $1 rpcbind_initrc_exec_t:file getattr;
+')
+
 ########################################
 ## <summary>
 ##	Search rpcbind lib directories.
diff --git a/policy/modules/system/systemd.te b/policy/modules/system/systemd.te
index 67ed3294b..2962e521e 100644
--- a/policy/modules/system/systemd.te
+++ b/policy/modules/system/systemd.te
@@ -2151,7 +2151,16 @@ userdom_user_runtime_root_filetrans_user_runtime(systemd_user_runtime_dir_t, dir
 userdom_manage_user_runtime_dirs(systemd_user_runtime_dir_t)
 userdom_mounton_user_runtime_dirs(systemd_user_runtime_dir_t)
 userdom_relabelto_user_runtime_dirs(systemd_user_runtime_dir_t)
-
+alsa_getattr_init_unit_file(systemd_generator_t)
+avahi_getattr_init_unit_file(systemd_generator_t)
+avahi_getattr_init_script_file(systemd_generator_t)
+bluetooth_getattr_init_unit_file(systemd_generator_t)
+bluetooth_getattr_init_script_file(systemd_generator_t)
+chronyd_getattr_init_unit_file(systemd_generator_t)
+chronyd_getattr_init_script_file(systemd_generator_t)
+rpcbind_getattr_init_script_file(systemd_generator_t)
+rpcd_getattr_init_unit_file(systemd_generator_t)
+init_read_generic_units_files(systemd_generator_t)
 optional_policy(`
 	dbus_system_bus_client(systemd_user_runtime_dir_t)
 ')
-- 
2.25.1
